pydoctor/epydoc/markup/epytext.py

```diff
@@ -382,23 +387,33 @@ def _pop_completed_blocks(token, stack, indent_stack):
             stack.pop()
             indent_stack.pop()
 
-def _add_para(doc, para_token, stack, indent_stack, errors):
+def _add_para(
+        para_token: 'Token',
+        stack: List[Element],
+        indent_stack: List[Optional[int]],
```

Comment:
> I guess that here None is used as `end of game` marker.
> In my code, I do something like `end_marker = object()` and then use end_marker instead of None.
> In this way, I find the code easier to read.

In this case, `None` is used for "indentation unknown", while -1 is used as a sentinel:
```py
    # Maintain two parallel stacks: one contains DOM elements, and
    # gives the ancestors of the current block.  The other contains
    # indentation values, and gives the indentation of the
    # corresponding DOM elements.  An indentation of "None" reflects
    # an unknown indentation.  However, the indentation must be
    # greater than, or greater than or equal to, the indentation of
    # the prior element (depending on what type of DOM element it
    # corresponds to).  No 2 consecutive indent_stack values will be
    # ever be "None."  Use initial dummy elements in the stack, so we
    # don't have to worry about bounds checking.
    stack = [cast(Element, None), doc]
    indent_stack = [-1, None]
```

> I guess that with typing, this adds extra complexity as you will first need to define a type.
> Or can you have `end_marker: EndMarkerT = object()` ?

Type names have to be defined as Python objects; you can't just introduce a name without defining it first.

Maybe [NewType](https://docs.python.org/3/library/typing.html?highlight=typing#typing.NewType) could be used for typed sentinels? I have never tried that though.

Reply:
Thanks for the explanation.

Thanks for the link. I think it will work and makes sense :)

```python
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0] on linux
>>> EndMarkerT = NewType('EndMarkerT', object)
>>> end_marker = EndMarkerT(object())
```

-------

What? NewType introduced in 3.5.**2** ?

Reply:
> What? NewType introduced in 3.5.**2** ?

There are a few more like that, for example `NoReturn`, which was new in 3.5.4 and 3.6.2. I guess that with type checking being a somewhat experimental feature at that time, they favored quick releasing over sticking to a "bug fixes only" policy.